feat(core): add option to infer the tag names of components in tests

crisbeto opened the pull request:
Currently when testing a component using `TestBed.createComponent`, we always create the component as a `div` which isn't aligned with the runtime. The runtime tries to parse out the tag name from the first selector in `@Component` and only falls back to `div` if there isn't one. This behavior difference can cause components to not behave like they would in production which reduces the usefulness of the tests.

These changes add the `inferTagName` option to `TestBed.createComponent` and `TestBed.configureTestingModule` that allows apps to opt into inferring the tag name from the selector in the same way as the runtime. Currently the new option is set to `false`, but we intend to change it to `true` in a future version.

jonathan-meier commented on packages/core/testing/src/test_bed.ts:
```diff
@@ -645,18 +646,24 @@ export class TestBedImpl implements TestBed {
+    const testComponentRenderer = this.inject(TestComponentRenderer);
+    const componentDef: ComponentDef<T> = (type as any).ɵcmp;
+    const rootElId = `root${_nextRootElementId++}`;
+    const defaultTag = componentDef.selectors[0][0] as string;
```
Does this also work for non-trivial selectors, e.g. [`button[matButton]`](https://github.com/angular/components/blob/4052130ee0dc627242a7f863f84eced4a0af048f/src/material/button/button.ts#L30)?

crisbeto replied:
Yes, this is the selector after it has been parsed by the compiler into a [`CssSelectorList`](https://github.com/angular/angular/blob/main/packages/core/src/render3/interfaces/projection.ts#L49)

pkozlowski-opensource replied:
@crisbeto it looks like it is roughly the same logic as in https://github.com/angular/angular/blob/b84859073bf094e0da5b509f393a738ad176edd9/packages/core/src/render3/component_ref.ts#L190-L192

The main difference being that component ref takes namespace into account (?). Maybe it would make sense to factor this out and re-use?

pkozlowski-opensource replied:
Yeh, I was just looking at it from the "don't duplicate code / concepts" side of things

thePunderWoman commented on packages/core/testing/src/test_bed.ts:
```diff
@@ -645,18 +646,24 @@ export class TestBedImpl implements TestBed {
+    // Note: injecting the renderer before accessing the definition appears to be load-bearing.
+    const testComponentRenderer = this.inject(TestComponentRenderer);
+    const componentDef: ComponentDef<T> = (type as any).ɵcmp;
+    const rootElId = `root${_nextRootElementId++}`;
```
Optional nit: It'd be nice to have a comment for the next two lines explaining what this is doing.

thePunderWoman reviewed (approved):
LGTM

reviewed-for: fw-general, public-api

devversion reviewed (approved):
LGTM

Reviewed-for: public-api

crisbeto committed:
refactor(core): expose function for inferring tag name

Exposes the code that we use to infer the tag name of a component as a private export so that we can use it in `TestBed`.

crisbeto committed:
feat(core): add option to infer the tag names of components in tests

Currently when testing a component using `TestBed.createComponent`, we always create the component as a `div` which isn't aligned with the runtime. The runtime tries to parse out the tag name from the first selector in `@Component` and only falls back to `div` if there isn't one. This behavior difference can cause components to not behave like they would in production which reduces the usefulness of the tests.

These changes add the `inferTagName` option to `TestBed.createComponent` and `TestBed.configureTestingModule` that allows apps to opt into inferring the tag name from the selector in the same way as the runtime. Currently the new option is set to `false`, but we intend to change it to `true` in a future version.
